### PR TITLE
Add mollie payments to invoices that are not yet paid in full

### PR DESCRIPTION
**Pull request by techhazard**

### CODE REVIEW

File: payment.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                // Renew payment
+            } else {
+                return payment_update;
+            }
+        }
+    })()
+
+
+    /*let payment;
+    if (paymentData !== null) {
```
Comment (jeroenpeeters) [inline payment.js]: remove commented out code

File: payment.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        const {payment_id} = paymentData;
+        console.log("payment exists: ", payment_id)
+        payment = await getPaymentUpdate(payment_id);
+    } else {
+        console.log("payment not found")
+        payment = await createPaymentData(invoice);
+    }*/
+    console.log("PAYMENT: ", payment);
```
Comment (jeroenpeeters) [inline payment.js]: remove or debug level

File: payment.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    });
+});
+
+
+
+exports.getPaymentForInvoice = (invoice) => new Promise( async (resolve, reject) => {
+
+    // look in DB for invoice_payment
+    let paymentData = await getPaymentFromDB(invoice.id).catch(null);
+    console.log("PAYMENTDATA: ", paymentData);
```
Comment (jeroenpeeters) [inline payment.js]: remove or debug level

File: payment.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+
+const getPaymentFromDB = (invoiceId) => new Promise( (resolve, reject) => {
+    console.log("looking for invoice: ", invoiceId)
+    payment_database.findOne({_id: invoiceId}, (err, invoice_payment) => {
+        if (err) {
+            console.log("Did not find invoice: ", err)
+            reject(err);
+        }
+        console.log("found invoice: ", invoice_payment)
```
Comment (jeroenpeeters) [inline payment.js]: remove or debug level

File: payment.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+
+const getPaymentUpdate = (payment_id) => mollie_client.payments.get(payment_id)
+
+
+const getPaymentFromDB = (invoiceId) => new Promise( (resolve, reject) => {
+    console.log("looking for invoice: ", invoiceId)
+    payment_database.findOne({_id: invoiceId}, (err, invoice_payment) => {
+        if (err) {
+            console.log("Did not find invoice: ", err)
```
Comment (jeroenpeeters) [inline payment.js]: remove or error level

File: payment.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    let value;
+    // if the string only has a single character after the dot, add a zero
+    // e.g. 10.0 -> 10.00, 12.5 -> 12.50
+    if (invoice.total_unpaid.match(/\..$/)) {
+        value = `${invoice.total_unpaid}0`;
+    } else {
+        value = invoice.total_unpaid;
+    }
+
+    if (value == '0.00' && MOLLIE_KEY.startsWith('test_')) {
```
Comment (jeroenpeeters) [inline payment.js]: don't write test variations in production code. if we cannot always connect to real Mollie, we should write a simple stub that implement the mollie API

File: moneybird.js [hunk 7]
```diff
@@ -1,21 +1,124 @@
...
-    return contact_request.body
+var fetch;
+switch (ENV) {
+    case 'prod':
+    case 'test':
+        fetch = new Fetch();
+        break;
+    case 'dev':
+    default:
+        fetch = new DevFetch();
```
Comment (jeroenpeeters) [inline moneybird.js:19]: What does this do exactly?

File: moneybird.js [hunk 8]
```diff
@@ -1,21 +1,124 @@
...
+var moneybird_invoices_cache = {};
+// the above, but indexed by invoice.id for easy lookup
+var moneybird_invoices_index = {};
+// contact data as received from moneybird
+var moneybird_contact_data_cache = {};
 
-    return invoices_request.body.filter( invoice => invoice.reference == reference )
+const getProjectName = (req) => {
+    if (ENV === 'dev') {
+        return '_development';
```
Comment (jeroenpeeters) [inline moneybird.js:35]: Same comment about test variations in production code. Do we really need this?

File: payment.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    const payment = await (async () => { 
+        if (paymentData === null) {
+            // no previous payment exists
+            return await createPaymentData(invoice);
+        } else {
+            // a previous payment exists, so we check it's status
+            const {payment_id} = paymentData;
+            const payment_update =await getPaymentUpdate(payment_id);
+            if (payment_update.status == 'expired') {
+                console.log("EXPIRED")
```
Comment (jeroenpeeters) [inline payment.js]: remove or debug. should we do anything when this happens?

**Review by jeroenpeeters — CHANGES_REQUESTED** [review]
Not sure if this can be resolved easily. But... in general you shouldn't distinguish between dev/test/prod in code. This makes code less reliable. This way code is programmed to behave differently in dev/test situations than production, which can yield some unexpected results. Besides, you need to maintain more code which makes it error prone.
I understand the need for variations in dev/test environments. Ideally you should solve this using a stub/mock implemented outside of the production codebase that implements the actual API.

If this is a lot of work we can go ahead for now but we need to fix this at a later point in time.

File: payment.js (hunk 1 above)
Comment (techhazard) [reply payment.js]: removed

File: payment.js (hunk 3 above)
Comment (techhazard) [reply payment.js]: removed

File: payment.js (hunk 5 above)
Comment (techhazard) [reply payment.js]: made into debug

File: payment.js (hunk 2 above)
Comment (techhazard) [reply payment.js]: removed

File: payment.js (hunk 4 above)
Comment (techhazard) [reply payment.js]: made into debug

File: moneybird.js (hunk 7 above)
Comment (techhazard) [reply moneybird.js:19]: I started with working in `server.js`, but after a while I decided to split code off to `moneybird.js`; I needed this as well, so I copied it from there.

As to what it does:
`Fetch` retrieves the variables needed from the client's files: `https://api.github.com/repos/${GITHUB_OWNER}/${GITHUB_REPO}/contents/${PROJECT_NAME}/${FILE_NAME}${ref_param}`, where `${FILE_NAME}` is often `subscrption.yml`
`DevFetch` retrieves the variables from `dev_config/subscription.yml` in the [hyperdev-io/subscription-api](https://github.com/hyperdev-io/subscription-api) repository instead.

File: moneybird.js (hunk 8 above)
Comment (techhazard) [reply moneybird.js:35]: While running in development mode, requests do not include a project name. 
This is a workaround, to make the code work even when no project name is available.

A better solution might be to always include a project name.

File: payment.js (hunk 9 above)
Comment (techhazard) [reply payment.js]: made into debug

Comment (techhazard) [reply payment.js]: We should, so I will

File: payment.js (hunk 6 above)
Comment (techhazard) [reply payment.js]: This was just convenient for testing, removed

**Comment by techhazard** [comment]
Wouldn't making a stub/mock just be moving the problem around?

Anyway, I updated the code, please review the changes :-)

File: moneybird.js (hunk 8 above)
Comment (techhazard) [reply moneybird.js:35]: something like "hyperdev-development", to prevent it from clashing with our clients

File: moneybird.js (hunk 7 above)
Comment (techhazard) [reply moneybird.js:19]: Please see `fetch.js` and `devFetch.js` for more context

**Pull request merged**